### PR TITLE
[GEARPUMP-23] add window dsl

### PR DESCRIPTION
**Pull request by manuzhang**
The PR is opened for early review and the work is in progress with following todos.

- [x] basic window dsl support with `WindowedWordCount` example
- [x] improve `ReduceFunction` to not emit intermediate results
- [x] add unit tests
- [ ] add comments and update documentation
- [ ] support different types of computation (e.g. monoid which doesn't require input elements to be held in the window)

### CODE REVIEW

File: streaming/src/main/scala/org/apache/gearpump/streaming/dsl/Stream.scala [hunk 1]
```diff
@@ -115,20 +121,17 @@ class Stream[T](
...
    */
-  def groupBy[Group](fun: T => Group, parallelism: Int = 1, description: String = null)
-    : Stream[T] = {
-    val groupOp = GroupByOp(fun, parallelism, Option(description).getOrElse("groupBy"))
-    graph.addVertex(groupOp)
-    graph.addEdge(thisNode, edge.getOrElse(Shuffle), groupOp)
-    new Stream[T](graph, groupOp)
+  def groupBy[GROUP](fn: T => GROUP, parallelism: Int = 1, description: String = "groupBy")
+    : GroupByStream[T, GROUP] = {
+    groupByWindow(DefaultGroupBy(fn), parallelism, description)
```
Comment (kkasravi) [inline streaming/src/main/scala/org/apache/gearpump/streaming/dsl/Stream.scala]: Seems odd: groupBy -> groupByWindow -> GroupByStream. Suggests that all groupBy's have windowing semantics.

File: streaming/src/main/scala/org/apache/gearpump/streaming/dsl/Stream.scala [hunk 2]
```diff
@@ -147,6 +150,33 @@ class Stream[T](
...
+    graph.addEdge(thisNode, edge.getOrElse(Shuffle), groupOp)
+    new GroupByStream[T, GROUP](graph, groupOp)
+  }
+
+}
+
+class GroupByStream[T, GROUP](
+    private val graph: Graph[Op, OpEdge],
+    private val groupByOp: GroupByOp[T, GROUP],
+    private val edge: Option[OpEdge] = None) extends Stream[T](graph, groupByOp, edge) {
```
Comment (kkasravi) [inline streaming/src/main/scala/org/apache/gearpump/streaming/dsl/Stream.scala]: Is 'private val graph' equivalent to 'graph' (val adds public accessors so private takes them away?)

File: streaming/src/main/scala/org/apache/gearpump/streaming/dsl/Stream.scala (hunk 1 above)
Comment (manuzhang) [inline streaming/src/main/scala/org/apache/gearpump/streaming/dsl/Stream.scala]: Yes, I have a new version that does so. Will update today

File: streaming/src/main/scala/org/apache/gearpump/streaming/dsl/Stream.scala (hunk 2 above)
Comment (manuzhang) [inline streaming/src/main/scala/org/apache/gearpump/streaming/dsl/Stream.scala]: `private val` is visible to companion object

**Comment by manuzhang** [comment]
@kkasravi 

`stream.groupBy` is now a shortcut for `stream.window(CountWindow.apply(1).triggering(CountTrigger).accumulating).groupBy`.

so `groupBy` has windowing semantics and all window operations should follow `groupBy`

here is the window semantics

```
- Window
    - WindowFn 
         -> SlidingWindowFn // FixedWindow and SlidingWindow 
         -> CountWindowFn  // CountWindow
    - Trigger
         -> EventTimeTrigger  // event time aggregation
         -> ProcessingTimeTrigger  // processing time aggregation
         -> CountTrigger // count aggregation, used with CountWindowFn
     - AccumulationMode
         -> Accumulating   // states are accumulated across windows
         -> Discarding  // states are not accumulated  
```

**Comment by kkasravi** [comment]
+1 let's get this merged and i'll rebase

**Pull request closed without merging**